File: KHost.App/ClientApp/src/app/modules/karaoke/providers/SongSearchProvider.ts
import { SongSearchEngine } from "../models/SongSearchEngine";
import { SongSearchResult } from "../models/SongSearchResult";

export class SongSearchProvider {

    constructor() {
        
    }

    search(query: string, songSearchEngine: SongSearchEngine, count: number = 20, offset: number = 0): Promise<SongSearchResult[]> {
        throw ("Not Implemented");
    }

    convertToLocalSong(songSearchResult: SongSearchResult): Promise<void> {
        throw ("Not Implemented");
    }

    convertToRemoteSong(songSearchResult: SongSearchResult): Promise<void> {
        throw ("Not Implemented");
    }

    getSongSearchEngines(): Promise<SongSearchEngine[]> {
        throw ("Not Implemented");
    }
}

File: KHost.App/Providers/SongSearchProvider.cs
﻿using KHost.Common.Models;
using KHost.App.Repositories;
using Microsoft.Extensions.DependencyInjection;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace KHost.App.Providers
{
    public class SongSearchProvider
    {
        private ISongSearchRepository[] SongSearchRepositories { get; }

        public SongSearchProvider(IServiceProvider services)
        {
            SongSearchRepositories = services.GetServices<ISongSearchRepository>().ToArray();
        }

        public async Task<IEnumerable<SongSearchResult>> Search(string searchQuery, string searchEngine, int? count = null, int? offset = null)
        {
            var repository = GetSongSearchRepositry(searchEngine);

            var songSearchResults = await repository.Search(searchQuery, count, offset);

            return songSearchResults;
        }

        public IEnumerable<SongSearchEngine> GetSongSearchEngineDefinitions()=> SongSearchRepositories.Select(r => r.EngineDefinition);

        private ISongSearchRepository GetSongSearchRepositry(string typeName)
        {
            foreach(var songSearchRepository in SongSearchRepositories)
            {
                if (songSearchRepository.GetType().Name == typeName)
                    return songSearchRepository;
            }

            return null;
        }
    }
}
